pkg/ccl/backupccl/incrementals.go

```diff
@@ -0,0 +1,168 @@
+	"github.com/cockroachdb/errors"
+)
+
+// The default subdirectory for incremental backups.
+const (
+	DefaultIncrementalsSubdir = "incrementals"
+	incBackupSubdirGlob       = "/[0-9]*/[0-9]*.[0-9][0-9]/"
+
+	// listingDelimDataSlash is used when listing to find backups and groups all the
+	// data sst files in each backup, which start with "data/", into a single result
+	// that can be skipped over quickly.
+	listingDelimDataSlash = "data/"
+)
+
+var BackupSubdirRE = regexp.MustCompile(`(.*)/([0-9]{4}/[0-9]{2}/[0-9]{2}-[0-9]{6}.[0-9]{2}/?)$`)
+
+// FindPriorBackups finds "appended" incremental backups by searching
+// for the subdirectories matching the naming pattern (e.g. YYMMDD/HHmmss.ss).
+// If includeManifest is true the returned paths are to the manifests for the
+// prior backup, otherwise it is just to the backup path.
+func FindPriorBackups(
+	ctx context.Context, store cloud.ExternalStorage, includeManifest bool,
+) ([]string, error) {
+	var prev []string
+	if err := store.List(ctx, "", listingDelimDataSlash, func(p string) error {
+		if ok, err := path.Match(incBackupSubdirGlob+backupManifestName, p); err != nil {
+			return err
+		} else if ok {
+			if !includeManifest {
+				p = strings.TrimSuffix(p, "/"+backupManifestName)
+			}
+			prev = append(prev, p)
+			return nil
+		}
+		if ok, err := path.Match(incBackupSubdirGlob+backupOldManifestName, p); err != nil {
+			return err
+		} else if ok {
+			if !includeManifest {
+				p = strings.TrimSuffix(p, "/"+backupOldManifestName)
+			}
+			prev = append(prev, p)
+		}
+		return nil
+	}); err != nil {
+		return nil, errors.Wrap(err, "reading previous backup layers")
+	}
+	sort.Strings(prev)
+	return prev, nil
+}
+
+func appendPaths(uris []string, tailDir ...string) ([]string, error) {
+	retval := make([]string, len(uris))
+	for i, uri := range uris {
+		parsed, err := url.Parse(uri)
+		if err != nil {
+			return nil, err
+		}
+		joinArgs := append([]string{parsed.Path}, tailDir...)
+		parsed.Path = JoinURLPath(joinArgs...)
+		retval[i] = parsed.String()
+	}
+	return retval, nil
+}
+
+// Stores are an odd combination of absolute and relative path.
+// They present as absolute paths, since they contain a hostname. URL.Parse
+// thus prepends each URL.Path with a leading slash.
+// But some schemes, e.g. nodelocal, can legally travel _above_ the ostensible
+// root (e.g. nodelocal://0/.../). This is not typically possible in file
+// paths, and the standard path package doesn't like it. Specifically, it will
+// clean up something like nodelocal://0/../ to nodelocal://0. This is normally
+// correct behavior, but is wrong here.
+//
+// In point of fact we block this behavior at other points. But we want to
+// make sure to resolve the paths correctly here. We don't want to accidentally
+// correct an unauthorized file path to an authorized one, then write a backup
+// to an unexpected place or print the wrong error message on a restore.
+func JoinURLPath(args ...string) string {
```

Comment:
i'm not totally clear on the subtle differences between `filepath.Join` and `path.join` and this helper function. Perhaps a few unit tests for this func would help illustrate? Maybe this function belongs in pkg/cloud/uris.go ? Perhaps @dt has more to say on this.

Reply:
Thanks david - good catch with the separator, I didn't think that through.

path.Join doesn't quite work - I tweaked the comments and added some unit tests. This any clearer?